### PR TITLE
Implementation of algorithm for finding full-width half-height markers for peaks.

### PR DESCRIPTION
**Pull request by Daroslaw**
Suggestions related to performance improvement are warmly welcomed.

This class will be used only for peak finding algorithms and is a part of larger pipeline, so for performance purposes the input data that goes into method of this class is assumed to be verified during earlier stages of the whole pipeline. More information in the methods documentation in header file.

### CODE REVIEW

**Comment by Daroslaw** [comment]
[Implement calculation of FWHH markers.](https://trello.com/c/TTSeenUc/402-implement-calculation-of-fwhh-markers)

File: src/Spectre.libPeakFinder.Tests/FWHHCalculatorTest.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    {
+    protected:
+        FWHHCalculator fwhh;
+        Signal x;
+
+        Signal sawtoothY;
+        Indices sawtoothValleyIdx;
+        Indices sawtoothPeakIdx;
+
+        void SetUp() override
```
Comment (Sand3r-) [inline src/Spectre.libPeakFinder.Tests/FWHHCalculatorTest.cpp]: Is it necessary to reassign the data upon each test? Perhaps we could move that to the test constructor.

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+#include "FWHHCalculator.h"
+#include "Spectre.libFunctional/Transform.h"
+
+namespace spectre::algorithm::peakfinder
+{
+    Signal FWHHCalculator::GetLeftFWHH(const SignalView x, const SignalView y, const IndicesView valleys, const IndicesView peaks)
+    {
+        Signal result;
+        result.reserve(peaks.size());
```
Comment (Sand3r-) [inline src/Spectre.libPeakFinder/FWHHCalculator.cpp]: It'd be more efficient to resize data and access by index instead of reserving, because on each push_back the length of vector needs to be increased and check for potential memory out-of-bounds write needs to be performed.

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            result.push_back(lfwhhVal);
+        }
+
+        return result;
+    }
+
+    Signal FWHHCalculator::GetRightFWHH(const SignalView x, const SignalView y, const IndicesView valleys, const IndicesView peaks)
+    {
+        Signal result;
+        result.reserve(peaks.size());
```
Comment (Sand3r-) [inline src/Spectre.libPeakFinder/FWHHCalculator.cpp]: Same here ofc.

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    inline Index FWHHCalculator::GetClosestNeighbourIndex(const SignalView & xSorted, DataType value)
+    {
+        Index valueIdx = 0;
+
+        for (unsigned i = 0; i < unsigned(xSorted.size()); ++i)
+        {
+            bool isGreater = xSorted[i] < value;
+            unsigned cond = unsigned(!isGreater) - 1;
+            i = i + (cond & xSorted.size());
+            valueIdx = (i - 1) & cond;
```
Comment (Sand3r-) [inline src/Spectre.libPeakFinder/FWHHCalculator.cpp]: It'd be nice to add comment what these 3 lines do, since it doesn't really come obvious from just casually reading that.

Comment (Sand3r-) [reply src/Spectre.libPeakFinder/FWHHCalculator.cpp]: I mean, even though I'm a co-author of part of that, I still don't quite get it :rofl:

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+        for (int i = 0; i < peaks.size(); ++i)
+        {
+            Index peakIdx = peaks[i];
+            Index valleyIdx = valleys[i + 1];
+            DataType peakVal = y[peakIdx];
+
+            SignalView segment = y.subspan(peakIdx, valleyIdx - peakIdx);
+            DataType rfwhhPeakVal = peakVal * 0.5;
+            Signal segmentInversed = spectre::core::functional::transform<DataType>(segment, [](DataType val) { return -val; });
```
Comment (Sand3r-) [inline src/Spectre.libPeakFinder/FWHHCalculator.cpp]: Rather segmentNegated ;)

**Review by Sand3r- — CHANGES_REQUESTED** [review]
Honestly a great read, most of the things come pretty clear, however there are a few changes to be considered.

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    inline Index FWHHCalculator::GetClosestNeighbourIndex(const SignalView & xSorted, DataType value)
+    {
+        Index valueIdx = 0;
+
+        for (unsigned i = 0; i < unsigned(xSorted.size()); ++i)
+        {
+            bool isGreater = xSorted[i] < value;
+            unsigned cond = unsigned(!isGreater) - 1;
+            valueIdx = (i - 1) & cond;
+            i = i + (cond & xSorted.size());
```
Comment (Sand3r-) [inline src/Spectre.libPeakFinder/FWHHCalculator.cpp]: Please, write a comment on what do those lines do.

**Review by Sand3r- — DISMISSED** [review]

File: src/Spectre.libPeakFinder.Tests/FWHHCalculatorTest.cpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    using namespace gsl;
+
+    class FWHHCalculatorTest : public Test
+    {
+    public:
+        FWHHCalculatorTest()
+        {
+            x = { 1., 2., 3., 4., 5., 6., 7., 8., 9. };
+
+            sawtoothY = { 0., 1., 2., 1., 0., 1., 2., 1., 0. };
```
Comment (gmrukwa) [inline src/Spectre.libPeakFinder.Tests/FWHHCalculatorTest.cpp:37]: What about additional case with a little bit skewed signal?

File: src/Spectre.libPeakFinder/FWHHCalculator.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        /// </summary>
+        /// <param name="x">Container view of m/z values.</param>
+        /// <param name="y">Container view of intensity values.</param>
+        /// <param name="valleys">Container view of valley indices.</param>
+        /// <param name="peaks">Container view of peak indices.</param>
+        /// <returns>New container of m/z values for right sides of FWHH markers for all peaks.</returns>
+        Signal GetRightFWHH(const SignalView x, const SignalView y, const IndicesView valleys, const IndicesView peaks);
+    private:
+        inline Index GetClosestNeighbourIndex(const SignalView& xSorted, const DataType value);
+        inline DataType Lerp(DataType x1, DataType x2, DataType y1, DataType y2, DataType x);   //TODO(dkuchta): Maybe in the future move it to libStatistics?
```
Comment (gmrukwa) [inline src/Spectre.libPeakFinder/FWHHCalculator.h]: Can we name it more expressively? Or at least add a docstring? It took me a longer while to figure out what's this.

**Review by gmrukwa — DISMISSED** [review]

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+#include "FWHHCalculator.h"
+#include "Spectre.libFunctional/Transform.h"
+
+namespace spectre::algorithm::peakfinder
+{
+    Signal FWHHCalculator::GetLeftFWHH(const SignalView x, const SignalView y, const IndicesView valleys, const IndicesView peaks)
+    {
+        Signal result;
+        result.resize(peaks.size());
```
Comment (gmrukwa) [inline src/Spectre.libPeakFinder/FWHHCalculator.cpp]: This can be done in single line.

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+        for (int i = 0; i < peaks.size(); ++i)
+        {
+            Index peakIdx = peaks[i];
```
Comment (gmrukwa) [inline src/Spectre.libPeakFinder/FWHHCalculator.cpp]: Most of these may be `const`.

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            Index valleyIdx = valleys[i];
+            DataType peakVal = y[peakIdx];
+
+            SignalView segment = y.subspan(valleyIdx, peakIdx - valleyIdx);
+            DataType lfwhhPeakVal = peakVal * 0.5;
+            Signal segmentNegated = spectre::core::functional::transform<DataType>(segment, [](DataType val) { return -val; });
+            Index lfwhhNearestIdx = GetClosestNeighbourIndex(segmentNegated, -lfwhhPeakVal) + valleyIdx;
+            DataType lfwhhVal = Lerp(y[lfwhhNearestIdx], y[lfwhhNearestIdx + 1], x[lfwhhNearestIdx], x[lfwhhNearestIdx + 1], lfwhhPeakVal);
+
+            result[i] = lfwhhVal;
```
Comment (gmrukwa) [inline src/Spectre.libPeakFinder/FWHHCalculator.cpp]: What about extracting calculation of single `result[i]` to an `inline` function of something and just running `transform`?

File: src/Spectre.libPeakFinder/FWHHCalculator.h [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        /// 5. Indices of both valleys and peaks are assumed to be reflecting real valleys/peaks.
+        /// </summary>
+        /// <param name="x">Container view of m/z values.</param>
+        /// <param name="y">Container view of intensity values.</param>
+        /// <param name="valleys">Container view of valley indices.</param>
+        /// <param name="peaks">Container view of peak indices.</param>
+        /// <returns>New container of m/z values for right sides of FWHH markers for all peaks.</returns>
+        Signal GetRightFWHH(const SignalView x, const SignalView y, const IndicesView valleys, const IndicesView peaks);
+    private:
+        inline Index GetClosestNeighbourIndex(const SignalView& xSorted, const DataType value);
```
Comment (gmrukwa) [inline src/Spectre.libPeakFinder/FWHHCalculator.h:69]: How the hell are you able to make `inline` function implementation in `.cpp` file instead of this one?

**Comment by gmrukwa** [comment]
[Investigate peak-finding algorithm](https://trello.com/c/UtEisfMo/187-investigate-peak-finding-algorithm)

File: src/Spectre.libPeakFinder/FWHHCalculator.h (hunk 12 above)
Comment (Sand3r-) [reply src/Spectre.libPeakFinder/FWHHCalculator.h:69]: https://stackoverflow.com/a/3993003
TL;DR Because the function is used only inside this cpp file.

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp (hunk 11 above)
Comment (Sand3r-) [reply src/Spectre.libPeakFinder/FWHHCalculator.cpp]: Totally agree with extracting. Especially since the code in GetRightFWHH is kind of a repetition of what's inside here.

**Comment by Sand3r-** [comment]
Wherever you use operator[] of gsl::span collection, please keep in mind that on each such access, the attempted index is checked against bounds of the array.

To mitigate this, you can:
1. Save the length to a constant
2. Cast the beginning of a span to pointer of underlying data type
3. Iterate over the array using the length and obtained pointer. 

This avoids unnecessary checks and going through unreasonable amount of branches through the program execution.

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp (hunk 11 above)
Comment (Daroslaw) [reply src/Spectre.libPeakFinder/FWHHCalculator.cpp]: This is very hard to do without doing lots of branching.

File: src/Spectre.libPeakFinder.Tests/FWHHCalculatorTest.cpp (hunk 7 above)
Comment (Daroslaw) [reply src/Spectre.libPeakFinder.Tests/FWHHCalculatorTest.cpp:37]: Done

**Comment by Daroslaw** [comment]
Holy moly @Sand3r- is soo right. Just look how long it takes to process `operator[]` of `gsl::span`.

gsljestfajny

**Comment by gmrukwa** [comment]
@Daroslaw, @Sand3r-, do you keep in mind, that these `constexpr` functions aren't function calls?

BTW are you sure, that these checks are also happening for `Release`? I know that MS implementation of STL has also a great number of checks & validators, that are just never used in `Release`, only in `Debug`. That's the reason TF works under VS in `Release` but in `Debug` doesn't.

Also, you shouldn't use pointer&size representation, that's what `gsl::span` is an abstraction of. Use `myspan.begin()` and `myspan.end()` if you want to go that way. Going to pointer&size we can throw away `span` in whole solution.

File: src/Spectre.libPeakFinder/FWHHCalculator.cpp (hunk 11 above)
Comment (gmrukwa) [reply src/Spectre.libPeakFinder/FWHHCalculator.cpp]: Well, no. You are checking only against the end of the collection. The rest of the code is linear. There is no branching.

EDIT: Now I get the point of separating the code repetition, that would introduce branching. If you even would go that way, you can make this function aggressively inlined or with `bool` as template parameter, so compiler will optimize away all `if`s.

Either way: readability over premature optimization. This one is premature. This code will not be ran soooo often.

**Review by gmrukwa — DISMISSED** [review]
Good job! BTW, I've updated: https://github.com/spectre-team/native-algorithms/pull/33#discussion_r173653031

**Comment by gmrukwa** [comment]
Guys, we're cluttering up the code with unpacking of `gsl::span`. Please wait with such premature optimization until we are able to profile the code.

**readability > premature optimization**

**Comment by Daroslaw** [comment]
Commit [`17651c0`](https://github.com/spectre-team/native-algorithms/commit/17651c095f4a11a57965aad5a6b528f7fa4d1057) nuked.

**Review by gmrukwa — APPROVED** [review]

**Comment by Sand3r-** [comment]
I did some profiling, and it looks like there is truly no performance hit. (At least on small chunks of data).
```
Run on (8 X 2594 MHz CPU s)
03/11/18 15:21:49
Benchmark           Time           CPU Iterations
-------------------------------------------------
BM_FWHHgsl        253 ns        248 ns    2639500
BM_FWHHraw        255 ns        254 ns    2639500
```

**Review by Sand3r- — APPROVED** [review]
